### ISSUE TITLE
Proposal: expose an API to get the origin of lines on a merge conflict.

### ISSUE THREAD

**Issue by cirosantilli** [libgit2/libgit2#2568]
Followup to: http://stackoverflow.com/questions/25849742/is-it-possible-to-get-the-origin-of-lines-on-a-diff3-merge-conflict-or-a-composi

I know how to generate a file containing the diff3: just copy the [following diff3 test](https://github.com/libgit2/libgit2/blob/3a495c19bd280f5455047a9ac0e936f2c2f2f9a9/tests/merge/workdir/simple.c#L259).

Suppose the diff3 would contain:

```
context1
<<<<<<< master
ours
||||||| base
base
=======
theirs
>>>>>>> topic
context2
```

What I want it to get the following map:

```
line        origin
-------------------
context1 -> CONTEXT
ours     -> OURS
base     -> BASE
theirs   -> THEIRS
context2 -> CONTEXT
```

possibly in a callback loop like is done for diff lines on [git_diff_print](https://libgit2.github.com/libgit2/#HEAD/group/diff/git_diff_print).

I want to do this because I am implementing a merge resolution tool for GitLab: https://github.com/gitlabhq/gitlabhq/pull/7345, so I need to break a diff3 into sections. (I have also proposed this for GitHub https://github.com/isaacs/github/issues/200)

Of course, I could parse the diff3 file and look for the conflict markers `<<<`, but that would lead to ambiguity if the file itself contained the markers, so I don't want to do that.

[ethomson has answered on SO](http://stackoverflow.com/a/25851717/895245) that what I want is currently not possible directly in libgit2, so my question is: would a pull request that allows this be considered, or is it out of scope?

I feel this may be in scope because libgit2 essentially already implements it to generate the conflict files: we would only be exposing what it does. Also the interface would look much like what `diff_print` currently does.

The simplest usage I can see is:
- operate on the index after a `git_merge`
- loop over conflict index entries with `index_conflict_next`, thus returning the ancestors, ours and theirs index entries for each file
- add the following interfaces:

``` c
// Mixture of index_print with merge_file.
int git_merge_file_from_index_foreach_line(
    git_repository *repo,
    const git_index_entry *ancestor,
    const git_index_entry *ours,
    const git_index_entry *theirs,
    const git_merge_file_options *opts,
    git_conflict_line_cb conflict_cb,
    void *payload
);

// Analogous to diff_line_cb.
typedef int (*git_conflict_line_cb)(
    const git_conflict_line *line,
    void *payload)

// Analogous to diff_line
typedef struct git_conflict_line {
    char   origin;                 /**< A git_conflict_line_t value */
    int    ancestor_lineno;   /**< Line number in ancestor file or -1 for added line */
    int    our_lineno;   /**< Line number in our file or -1 for deleted line */
    int    their_lineno;   /**< Line number in their file or -1 for deleted line */
    int    num_lines;    /**< Number of newline characters in content */
    size_t content_len;  /**< Number of bytes of data */
    git_off_t content_offset; /**< Offset in the original file to the content */
    const char *content; /**< Pointer to diff text, not NUL-byte terminated */
} git_conflict_line;

// Analogous to diff_line_t
typedef enum {
    /* THIS is what I'm after. */
    GIT_CONFLICT_LINE_CONTEXT   = ' ',
    GIT_CONFLICT_LINE_OUR  = 'o',
    GIT_CONFLICT_LINE_THEIRS  = 't',
    GIT_CONFLICT_LINE_BOTH  = 'b',

    // TODO: think about the possible EOF cases.
    GIT_CONFLICT_LINE_CONTEXT_EOFNL = '=', /**< Both files have no LF at end */
    GIT_CONFLICT_LINE_ADD_EOFNL = '>',     /**< Old has no LF at end, new does */
    GIT_CONFLICT_LINE_DEL_EOFNL = '<',     /**< Old has LF at end, new does not */
} git_conflict_line_t;

```

Other options which would also work for me include:
- offer a line-wise callback combined diff interface (the default diff format for merge conflicts). Would loop only over lines around the conflicts.
- offer a line-wise callback interface that does not require a previous merge, since it is a bit redundant merging twice.

**Comment by tarrant** [comment]
So I was recently bit by this missing feature.

My first path was to write code against upstream libxdiff using `xdl_merge3` this got me what I wanted but when I then went and tried to compile this with libgit2 I ran into an unpleasant surprise.

It turns out libgit2 has forked xdiff from git/git which forked it from upstream x-mailserver. The feature `xdl_merge3` post-dates the first of these forks. This means that the function is unavailable. In addition the libraries have diverged to such a degree that you can't easily compile both libxdiff (upstream) along side libgit2.

If there is interest I'm interested in looking into either writing a custom merge function in the forked xdiff library or porting xdl_merge3 from upstream into libgit2 and adding bindings. But if there is no chance of getting these patches merged then I'll follow a different path.

**Comment by ethomson** [comment]
We try to track the version of xdiff that's in [git.git](https://github.com/git/git), not necessarily the upstream (x-mailserver) version, because we share more in common with that version than the upstream (as you might expect).

My first question would be:  does git.git include `xdl_merge3` in their xdiff?  If so, then the course of action would be to update our version based on that and go from there.

**Comment by tarrant** [comment]
Looks like git.git doesn't include xdl_merge3 either.

**Comment by ethomson** [comment]
/cc @probablycorey - this may be relative to your interests as well.

I would say that the next step, then, depends on what git.git has planned for xdiff.  If they're amenable to taking an update to it, then we would certainly be willing to match that update.  @peff - could you shed any light on this?

**Comment by peff** [comment]
We don't have any plans for xdiff. I don't think we'd be against updating our version of xdiff if it brought in new features or bugfixes. And maybe even we'd be for it in the sense of just keeping our fork as close to the original as possible, which would make future features and bugfixes easier to share. But it would first take somebody figuring out what the changes _are_, and volunteering to merge them in.

We could also potentially just take the `xdl_merge3` bits. But if nobody in git.git is going to call them, then I"m not sure we're especially interested. I'm not even clear on what `xdl_merge3` is supposed to do versus `xdl_merge`. It looks like the latter is not even present in libxdiff-0.23.

It's not even clear to me what changes have been made since libxdiff was pulled into git, or even which version of libxdiff was pulled in, or what changes were made to it. xdiff seems to have no version control, and the changelog has a single entry which is clearly related to Linus's initial complaint from git/git@3443546f6ef57fe28ea5cca232df8e400bfc3883.

So...yeah. I don't even think I can comment on what upstream git would be interested in doing until somebody teases apart just what the proposed changes would be. And I do not want that somebody to be me. :)

**Comment by ethomson** [comment]
Right, my understanding here is that `xdl_merge` was dropped in favor of `xdl_merge3`.  It looks to me, from the documentation, like the new `xdl_merge3` is more powerful and can provide common / conflict / etc line numbers and such.  Which is very handy for the proposed issue, but it seems to make things a lot less easy to use (for us) since now we have to write the decorations ourselves.  But don't treat that assessment as gospel, as I literally spent 10 minutes looking at this before deciding this problem to be hard.

I don't want to deadlock this issue by having upstream git and us pointing our guns at each other and suggesting that we won't budge until the other does first.  But I did want to set expectations appropriately.

Certainly, @tarrant I think that we'd all be interested in seeing how costly an upgrade would be...!

**Comment by peff** [comment]
In case anybody is interested in going further: digging on the mailing list, it looks like libxdiff-0.17 was what got pulled in to git. Linus claims:

> I'd be very surprised if we have any differences in the diff output wrt libxdiff-0.17. I was really pretty careful, and didn't change the code at all, just removed unnecessary files and functions.

So in theory one could make a "vendor branch" by checking out commit 3443546, copying xdiff-0.23 on top of it, and re-stripping out the Linus bits. And then merge that up, and see if it conflicts with any of the changes we've made outside of libxdiff.
